bitcoin/src/crypto/key.rs

```diff
@@ -863,76 +861,150 @@ impl From<TweakedKeypair> for TweakedPublicKey {
     fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
-        use Error::*;
+        use FromSliceError::*;
 
         match *self {
-            Base58(ref e) => Some(e),
             Secp256k1(ref e) => Some(e),
-            Hex(ref e) => Some(e),
-            InvalidKeyPrefix(_) | InvalidHexLength(_) => None,
+            InvalidKeyPrefix(_) | InvalidLength(_) => None,
         }
     }
 }
 
-impl From<base58::Error> for Error {
-    fn from(e: base58::Error) -> Error { Error::Base58(e) }
+impl From<secp256k1::Error> for FromSliceError {
+    fn from(e: secp256k1::Error) -> FromSliceError { Self::Secp256k1(e) }
+}
+
+/// Error generated from WIF key format.
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[non_exhaustive]
+pub enum FromWifError {
+    /// A base58 error.
+    Base58(base58::Error),
+    /// A secp256k1 error.
+    Secp256k1(secp256k1::Error),
+}
+
+impl fmt::Display for FromWifError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use FromWifError::*;
+        match *self {
+            Base58(ref e) => write_err!(f, "invalid base58"; e),
+            Secp256k1(ref e) => write_err!(f, "private key validation failed"; e),
+        }
+    }
+}
+
+impl From<base58::Error> for FromWifError {
+    fn from(e: base58::Error) -> FromWifError { Self::Base58(e) }
+}
+
+impl From<secp256k1::Error> for FromWifError {
+    fn from(e: secp256k1::Error) -> FromWifError { Self::Secp256k1(e) }
+}
+
+/// Error returned while constructing public key from string.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum ParsePublicKeyError {
+    /// Error originated while parsing string.
+    Encoding(FromSliceError),
+    /// Hex decoding error.
+    InvalidChar(u8),
+    /// `PublicKey` hex should be 66 or 130 digits long.
+    InvalidHexLength(usize),
+}
+
+impl fmt::Display for ParsePublicKeyError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use ParsePublicKeyError::*;
+        match *self {
+            Encoding(ref e) => write_err!(f, "string error"; e),
+            InvalidChar(ref char) => write_err!(f, "hex error"; char),
+            InvalidHexLength(ref got) => write!(f, "pubkey string should be 66 or 130 digits long, got: {}", got),
+        }
+    }
+}
+
+impl From<hex::HexToArrayError> for ParsePublicKeyError {
+    fn from(e: hex::HexToArrayError) -> ParsePublicKeyError {
+        match e {
+            hex::HexToArrayError::Conversion(HexToBytesError::InvalidChar(char)) => Self::InvalidChar(char),
+            _ => panic!("This should have never happened"),
+        }
+    }
+}
+
+impl From<FromSliceError> for ParsePublicKeyError {
+    fn from(e: FromSliceError) -> Self { Self::Encoding(e) }
+}
+
+/// Error returned when parsing a [`CompressedPublicKey`] from a string.
+#[derive(Debug)]
+pub enum ParseCompressedPublicKeyError {
+    /// Secp256k1 Error.
+    Secp256k1Error(secp256k1::Error),
+    /// hex to array conversion error.
+    HexError(hex::HexToArrayError),
```

Comment:
Variants should not include `Error` in the name. These should be
```rust
   /// Secp256k1 error deserializing key.
   Secp256k1(secp256k1::Error),
   /// Hex conversion error.
   Hex(hex::HexToArrayError),
```

Reply:
Sorry I did not notice this in earlier reviews.

Reply:
I think you did. I forgot to change it here. Thanks.